### ISSUE TITLE
7.0.3 introduces breaking change to `touch` and `update_columns`

### ISSUE THREAD

**Issue by mroach** [rails/rails#45054]
### Steps to reproduce

This test works with `7.0.2` but fails with `7.0.3`:

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails", "7.0.3"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.datetime :archived_at
    t.timestamps
  end

  create_table :comments, force: true do |t|
    t.integer :post_id
  end
end

class Post < ActiveRecord::Base
  has_many :comments

  def readonly?
    persisted? && archived_at?
  end
end

class Comment < ActiveRecord::Base
  belongs_to :post, touch: true
end

class BugTest < Minitest::Test
  def test_unarchiving_post
    post = Post.create!(archived_at: Time.now)
    post.update_columns(archived_at: nil)

    assert_nil post.archived_at
  end

  def test_touching_archived_post
    post = Post.create!(archived_at: Time.now)
    post.comments.create!

    assert_equal 1, post.comments.count
  end
end
```

### Expected behavior

`update_columns` and `touch` continue to work on `readonly?` records at least until the next major version. If the behaviour is to change, a deprecation warning for using these methods on `readonly?` records would be desirable.

### Actual behavior

`update_columns` and `touch` now raise `ActiveRecord::ReadOnlyRecord` when the record is readonly.

### System configuration
**Rails version**: `7.0.3`

**Ruby version**: `3.1.1`

### Probable cause

https://github.com/rails/rails/pull/44845

**Comment by eileencodes** [comment]
Why would it be desirable for a readonly record to get updated? https://github.com/rails/rails/pull/44845 was to fix a bug and having bug fixes in a point release is fine.

**Comment by eileencodes** [comment]
I saw your tweet and I'm not going to discuss this on twitter because it's too hard to write code etc. We don't normally add bugfixes to the changelog, which is why it's not there. It's too late to add it now, the release is already done.

If you want to update a record you should mark it as not readonly first. You can do that with `post.readonly(false)`.

**Comment by caseyprovost** [comment]
FWIW it appears this change has also broke our test suite. We do a lot of readonly object updating in test and that has worked consistently for years (since at least rails 5.2). It would be nice if we could revert, deprecate, and fix in a minor version but totally understand if that's not possible.

**Comment by mroach** [comment]
> Why would it be desirable for a readonly record to get updated and not raise? It's readonly, so it shouldn't allow any updates through.

`update_columns` has been the go-to for writing values to the database while bypassing validations, callbacks, and until now, the `readonly?` check. This was one step away from writing raw SQL. In the given example where the `readonly?` status of a record is determined by data, raw SQL would appear to now be the only way to undo that.

I can understand an argument for taking readonly into consideration and I think that makes sense. But that's a pretty big change in behaviour that's been around for years if not always. The [documentation for `update_columns`](https://api.rubyonrails.org/v7.0.3/classes/ActiveRecord/Persistence.html#method-i-update_columns) also does not mention this new restriction that will raise a `ActiveRecord::ReadOnlyRecord`.

>  #44845 was to fix a bug and having bug fixes in a point release is fine.

I had thought, perhaps incorrectly, that Rails was following Semantic Versioning where breaking backwards compatibility was reserved for major versions.

> We don't normally add bugfixes to the changelog, which is why it's not there. 

There are a dozen fixes noted in the current [ActiveRecord changelog](https://github.com/rails/rails/blob/main/activerecord/CHANGELOG.md) and this change breaks backwards compatibility, so a mention of it would have been handy to figure out failing builds.

> If you want to update a record you should mark it as not readonly first. You can do that with post.readonly(false).

This doesn't appear to work in the context of a single object. There's `readonly!` but that can only switch it on.

```
> u = User.first
> u.readonly(false)
/usr/local/bundle/gems/activemodel-7.0.3/lib/active_model/attribute_methods.rb:458:in `method_missing': undefined method `readonly' for #<User id: 1...
```

**Comment by matthewd** [comment]
> I had thought, perhaps incorrectly, that Rails was following Semantic Versioning where breaking backwards compatibility was reserved for major versions.

Yeah, no, we do not follow semantic versioning; three-part version numbers are much older than that particular interpretation of them. We regularly deprecate and then remove chunks of functionality in sequential minor releases.

All bug fixes break backwards compatibility with code that was depending on the presence of the bug. For truly perfect compatibility, the only option is to not update at all.

> The [documentation for update_columns](https://api.rubyonrails.org/v7.0.3/classes/ActiveRecord/Persistence.html#method-i-update_columns) also does not mention this new restriction that will raise a ActiveRecord::ReadOnlyRecord.

It does say "or when at least one of the attributes is marked as readonly", though doesn't explicitly call out the possibility that the whole record might be readonly. But then, nor does `save`... and `delete` does explicitly carve itself out as an exception because it _does_ affect `readonly?` records. Here's what `readonly` says, though (emphasis added):

> Sets readonly attributes for the returned relation. If value is true (default), **attempting to update a record will result in an error**.

That seems like a reasonably clear statement of intended behaviour to me?

> Edit: update_attribute and update permit updating the column that's used in the readonly? check. That's unexpected.

So too does.. just setting the attribute: `post.archived_at = nil; post.save!`.

As far as I'm aware, overriding `readonly?` is undefined behaviour. If you're going to do that, you could also just add a check for `@actually_lets_allow_writing_after_all` to the conditional... but either way, I would discourage leaning on "if I monkey-patch the right method, I can create an inconsistency" as an argument.

---

If you want an option that doesn't modify your existing model definition, `Relation#update_all` is documented as bypassing model instantiation, and therefore will not have an opportunity to encounter your hacked `readonly?`. While raw SQL is a perfectly reasonable tool to achieve a variety of goals, it's not necessary here.

**Comment by caseyprovost** [comment]
@matthewd thanks for that explanation! This has brought to light some bad practices we were following and motivates us to fix that :) Appreciate your care and thoughtful approach

**Comment by mroach** [comment]
> As far as I'm aware, overriding `readonly?` is undefined behaviour. 

This is the crux of the matter and source of my issue. Trying to extend/modify/piggy-back on the standard readonly functionality seems to be the wrong approach here and I'll have to refactor away from doing that. `update_all` may be a good temporary workaround to unblock upgrading; nice suggestion!

I appreciate your spending the time to respond and work through this.

**Comment by eileencodes** [comment]
Since this is not a bug and is intended behavior I'm going to close this.
